hil/ext/switches/juniper.py

```diff
@@ -0,0 +1,373 @@
+
+# Copyright 2013-2017 Massachusetts Open Cloud Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+"""A switch driver for the juniper QFX series.
+
+But in theory can work for any Juniper switch that supports NETCONF.
+Driver uses the python library PyEZ made available by Juniper.
+By default it sets NETCONF session over ssh, i.e. make RPC calls over ssh.
+For this to work, make sure that the Juniper switch is configured
+appropriately to recieve NETCONF connections. Its default port is 830.
+"""
+
+import re
+import logging
+import schema
+import json
+import sys
+import os
+
+from jnpr.junos import Device
+from jnpr.junos.exception import ConnectError
+from jnpr.junos.utils.config import Config
+from jnpr.junos.utils.config import ConfigLoadError
+
+
+from hil.model import db, Switch, Network
+from hil.model import BigIntegerType
+from hil.migrations import paths
+from os.path import dirname, join
+from hil.migrations import paths
+from hil.ext.switches.junos.config_tables.ConfigTables import (
+        InterfaceConfigTable
+        )
+logger = logging.getLogger(__name__)
+paths[__name__] = join(dirname(__file__), 'migrations', 'juniper')
+
+
+class VlanAddError(Exception):
+    pass
+
+
+class ConfigCommitError(Exception):
+    pass
+
+
+class Juniper(Switch):
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/juniper'
+    dir_name = os.path.dirname(__file__)
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(
+            BigIntegerType, db.ForeignKey('switch.id'), primary_key=True
+            )
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+
+    enable_interface = None
+    add_vlan = None
+    remove_vlan = None
+    set_native_vlan = None
+    remove_native_vlan = None
+    remove_default_vlan = None
+    set_trunk_mode = None
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'username': basestring,
+            'hostname': basestring,
+            'password': basestring,
+        }).validate(kwargs)
+
+    def session(self):
+        return self
+
+    def disconnect(self):
+        pass
+
+    def _create_session(self):
```

Comment:
why did you make a different session object here instead of at line 87 `session`?

Reply:
if I create the session object, I have to put all the other function under the session method and that was too much indentation for me. 
To get some more horizontal space I reduced a level of identation by not using the `session` object. Instead made a private method.

Also, it only creates the object. Opening, loading, closing the session is done whenever a method requires the session.

Reply:
You can solve the indentation problem by declaring a wrapper class elsewhere in the file, and here just return that.